sympy/codegen/rewriting.py

```diff
@@ -175,42 +175,99 @@ def optimize(expr, optimizations):
 )
 
 
-class _FuncMinusOne:
-    def __init__(self, func, func_m_1):
+class FuncMinusOneOptim(ReplaceOptim):
+    """Specialization of ReplaceOptim for functions evaluating "f(x) - 1".
+
+    Explanation
+    ===========
+
+    Numerical functions which go toward one as x go toward zero is often best
+    implemented by a dedicated function in order to avoid catastrophic
+    cancellation. One such example is ``expm1(x)`` in the C standard library
+    which evaluates ``exp(x) - 1``. Such functions preserves many more
+    significant digits when its argument is much smaller than one, compared
+    to subtracting one afterwards.
+
+    Parameters
+    ==========
+
+    func :
+        The function which is subtracted by one.
+    func_m_1 :
+        The specialized function evaluating ``func(x) - 1``.
+    opportunistic : bool
+        When ``True``, apply the transformation as long as the magnitude of the
+        remaining number terms decreases. When ``False``, only apply the
+        transformation if it completely eliminates the number term.
+
+    Examples
+    ========
+
+    >>> from sympy import symbols, exp
+    >>> from sympy.codegen.rewriting import FuncMinusOneOptim
+    >>> from sympy.codegen.cfunctions import expm1
+    >>> x, y = symbols('x y')
+    >>> expm1_opt = FuncMinusOneOptim(exp, expm1)
+    >>> expm1_opt(exp(x) + 2*exp(5*y) - 3)
+    expm1(x) + 2*expm1(5*y)
+
+
+    """
+
+    def __init__(self, func, func_m_1, opportunistic=True):
+        weight = 10  # <-- this is an arbitrary number (heuristic)
+        super().__init__(lambda e: e.is_Add, self.replace_in_Add,
+                         cost_function=lambda expr: expr.count_ops() - weight*expr.count(func_m_1))
         self.func = func
         self.func_m_1 = func_m_1
+        self.opportunistic = opportunistic
 
-    def _try_func_m_1(self, expr):
-        old_new = {}
-        protected = expr.replace(self.func, lambda arg: old_new.setdefault(arg, Dummy()))
-        factored = protected.factor()
-        new_old = {v: self.func(k) for k, v in old_new.items()}
-        return factored.replace(_d - 1, lambda d: self.func_m_1(new_old[d].args[0])).xreplace(new_old)
-
-    def __call__(self, e):
-        numbers, non_num = sift(e.args, lambda arg: arg.is_number, binary=True)
-        non_num_func, non_num_other = sift(non_num, lambda arg: arg.has(self.func),
-            binary=True)
+    def _group_Add_terms(self, add):
+        numbers, non_num = sift(add.args, lambda arg: arg.is_number, binary=True)
```

Comment:
I wonder if you should further sift numbers into Rational and other:
```python
>>> optimize(2*exp(x)-2,[expm1_opt]).subs(x,1e-11)
2.00000000001000e-11
>>> (2*exp(x)-2).subs(x,1e-11)
2.00000016548074e-11

>>> optimize(2*exp(x)-pi,[expm1_opt])
2*expm1(x) - pi + 2
>>> _.subs(x,1e-11)
2.00000000002 - pi
>>> (2*exp(x)-pi).subs(x,1e-11)
2.00000000002 - pi
```

Reply:
Ah, yes, not very beautiful, but I am not sure how `is_rational` is central to avoiding cancellation? For example, if I carefully construct a fraction of pi close to two:
```python
In [31]: about_two = 2**(58/S(117))*3**(97/S(117))*5**(4/S(39))*7**(92/S(117))/S(30)*pi

In [32]: xval = S(10)**-11

In [33]: ref = (2*exp(x) - about_two).subs(x, xval).n(17); print(ref)
2.0418173913673214e-11

In [34]: no_opt = (2*exp(x) - about_two).subs(x, xval.evalf()).evalf()

In [35]: with_opt = optimize(2*exp(x) - about_two, [expm1_opt]).n(25).subs(x, 1e-11).evalf()

In [36]: with_opt - ref, no_opt - ref
Out[36]: (1.1536301877952077e-26, 1.6547074214222335e-18)
```
somewhat orthogonal to what we are discussing: note how I needed to insert `.n(25)` there, `evalf` didn't cut it. Hopefully our (code) printers compute such sums to the full precision of the targeted data type... (but this is another issue of course).

Reply:
What the example I gave might be saying is that there is no benefit to rewriting unless you can absorb an integer; if the number is not rational then it looks like there is no benefit. 

As far as the examples go...I'm nervous about the multiple evalfs that you use to demonstrate the difference. I need a more direct demonstration that the optimization is more precise.

Reply:
Right, I wanted to make sure myself, so I did spend some time adding a test which does a full roundtrip to (generated) C code and back. I just added those commits here.